x/ecocredit/server/marketplace/utils.go

```diff
@@ -18,3 +25,186 @@ func isDenomAllowed(ctx sdk.Context, denom string, pk ecocredit.ParamKeeper) boo
+				Tradable: "0",
+				Retired:  "0",
+				Escrowed: "0",
+			}
+		} else {
+			return err
+		}
+	}
+	if !opts.autoRetire { // if auto retire is disabled, we move the credits into the buyer's/supply's tradable balance.
+		tradableBalance, err := math.NewDecFromString(buyerBal.Tradable)
+		if err != nil {
+			return err
+		}
+		tradableBalance, err = math.SafeAddBalance(tradableBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Tradable = tradableBalance.String()
+
+		supplyTradable, err := math.NewDecFromString(supply.TradableAmount)
+		if err != nil {
+			return err
+		}
+		supplyTradable, err = math.SafeAddBalance(supplyTradable, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.TradableAmount = supplyTradable.String()
+		if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventReceive{
+			Sender:         sdk.AccAddress(sellOrder.Seller).String(),
+			Recipient:      buyerAcc.String(),
+			BatchDenom:     opts.batchDenom,
+			TradableAmount: purchaseQty.String(),
+			RetiredAmount:  "",
+			BasketDenom:    "",
+		}); err != nil {
+			return err
+		}
+	} else {
+		retiredBalance, err := math.NewDecFromString(buyerBal.Retired)
+		if err != nil {
+			return err
+		}
+		retiredBalance, err = math.SafeAddBalance(retiredBalance, purchaseQty)
+		if err != nil {
+			return err
+		}
+		buyerBal.Retired = retiredBalance.String()
+
+		supplyRetired, err := math.NewDecFromString(supply.RetiredAmount)
+		if err != nil {
+			return err
+		}
+		supplyRetired, err = math.SafeAddBalance(supplyRetired, purchaseQty)
+		if err != nil {
+			return err
+		}
+		supply.RetiredAmount = supplyRetired.String()
+		if err = sdkCtx.EventManager().EmitTypedEvent(&core.EventRetire{
+			Retirer:    buyerAcc.String(),
+			BatchDenom: opts.batchDenom,
+			Amount:     purchaseQty.String(),
+			Location:   opts.retirementLocation,
+		}); err != nil {
+			return err
+		}
+	}
+	if err = k.coreStore.BatchSupplyTable().Update(ctx, supply); err != nil {
+		return err
+	}
+	if err = k.coreStore.BatchBalanceTable().Save(ctx, buyerBal); err != nil {
+		return err
+	}
+
+	return k.bankKeeper.SendCoins(sdkCtx, buyerAcc, sellOrder.Seller, sdk.NewCoins(cost))
+}
+
+// getTotalCost calculates the cost of the order by multiplying the price per credit, and the amount of credits
+// desired in the order.
+func getTotalCost(pricePerCredit sdk.Int, amtCredits math.Dec) (sdk.Int, error) {
+	unitPrice, err := math.NewPositiveFixedDecFromString(pricePerCredit.String(), amtCredits.NumDecimalPlaces())
+	if err != nil {
+		return sdk.Int{}, err
+	}
+	cost, err := amtCredits.Mul(unitPrice)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	// TODO: update with Dec ToInt function #987
+	// divide by 1 to clear anything to the right of the decimal, so we can convert to sdk.Int
+	cost, err = cost.QuoInteger(math.NewDecFromInt64(1))
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	costStr := cost.String()
+
+	costInt, ok := sdk.NewIntFromString(costStr)
```

Comment:
I just found that we have `BigInt` function - so instead of doing a roundtrip through string, let's use `big.Int` -> `sdk.Int`

Reply:
https://github.com/regen-network/regen-ledger/pull/967/commits/ae9ef5995108e1a7484dcfbaf0a147cd8ff2865e